### ISSUE TITLE
earliest/latest_by_offset should not ignore null values

### ISSUE THREAD

**Issue by big-andy-coates** [confluentinc/ksql#5727]
I don’t think `LATEST_BY_VALUE` and `EARLIEST_BY_VALUE` should skip nulls.  Skipping nulls is a bug as far as I’m concerned.

Consider a stream of facts:

```
Key: A, Col0: 10
Key: A, Col0: null
Key: A, Col0: 20
```

Currently LATEST_BY_OFFSET(Col0) will output the sequence `10`, `10`, `20`, when clearly the values should be `10`, `null`, `20`.  A `null` value in a column should hold no special meaning for this UDAF and should not be ignored.

(Raised of the back of several users confused by current behaviour)

**Comment by purplefox** [comment]
This is by design, please see the comment on the original PR:

https://github.com/confluentinc/ksql/pull/4782

"Imo for the kinds of use cases that this is intended for... stock price updates, IoT sensor updates I suspect ignoring nulls is ok for now (most probably they won't send nulls). I'd suggest going with this for now and creating a new variant which doesn't ignore nulls if there is user demand for it."

TLDR; some people will want nulls other won't. The current version ignores nulls, but there's certainly scope for another version which doesn't ignore nulls.

**Comment by big-andy-coates** [comment]
Users are reporting this isn't what they need, (though arguably there's likely to be another set of users where it does what they need).

Personally, I don't think we should treat NULL as a special column value by default. The current null handling may be intentional, but my own view would be to have these methods treat null like any other value.   This then means these functions won't work for spare-data use-cases, i.e. where each row contains only values for some of the columns and users won't to only update non-null columns.  This feels like a more niche use-case to me.  We could add `SPARSE_LATEST_BY_OFFSET` for this use-case.

Of course, another perfectly valid view of this is to say that this is not a bug, as it was by design and inline with the docs. So if we want a version that doesn't treat nulls specially, then we could add that... just not sure of a nice intuitive name for such functions... any ideas?

The nicest solution would be to have these methods take an optional second parameter to control null handling, e.g. a `ignoreNulls` boolean or similar.  Unfortunately, like may UDAF fixes and enhancements, this requires [enhancements to the UDAF framework to allow additional parameters](https://github.com/confluentinc/ksql/issues/5747), which isn't getting any priority at the moment. This would likely result in the least disruptive short term, and intuitive long term, solution.

**Comment by mjsax** [comment]
Can we please clarify if this discussion is about Java `null` values or SQL `NULL` values? Note, for Java `null` values Kafka Streams treats them as invalid input to an aggregation (what seems sounds) and skips them -- thus, ksqlDB would never seem them to begin with. For SQL `NULL` handling, it's a different story though.

**Comment by big-andy-coates** [comment]
@mjsax I don't follow. While Kafka Streams doesn't allow `null`s as a valid aggregation `key` it enforces no restrictions on `null`s in the fields of the value, or indeed on the value itself being `null`. At least none that I'm aware of.

What we're talking about here is passing a source column, (or indeed some other expression), to a UDAF that is null: both SQL `NULL` and Java `null`.  We are _not_ talking about aggregating/grouping by a `null` or `NULL` key. We're talking about grouping by some non-null field of the data and passing another possibly-null field into a UDAF.

**Comment by mjsax** [comment]
> it enforces no restrictions on nulls in the fields of the value, or indeed on the value itself being null. At least none that I'm aware of.

If does: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamAggregate.java#L84-L92

Not sure how ksqlDB treats `null` values? If if keeps them as-is, KS would drop those records in an aggregation step. Of ksqlDB translates them to a "record" with all "value columns" being set to `NULL`, the underlying Java value would not be `null` and thus those `NULL`s are processed by KS.

**Comment by big-andy-coates** [comment]
> It does: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamAggregate.java#L84-L92

Good to know.

So full `null` value would currently be ignored/dropped, where as a value with all `null` fields would be processed and pass `null` to the UDAF.

Obviously, for tables a `null` value _is_ special, i.e. a tombstone.  For streams it really isn't.  I wonder if ksqlDB should map a `null` value into a structure with all null columns....   (Off topic for this issue though)

**Comment by mjsax** [comment]
The root cause is left/outer stream-stream joins. For both operators, if we get input `<k,null>` and the `ValueJoiner` would be called, the left/right value of `apply(V1 left, V2 right)` might be `null` -- and this would introduce an ambiguity between a `<k,null>` input and if a key does **_not_** match (ie, a left/right outer call of `ValueJoiner#apply`). To avoid the ambiguity we decided that `<k,null>` is considered "malformed" as "there is nothing to be joined" anyway. -- To be consistent across all DSL operators, we apply the same logic (ie, "there is nothing to be aggregated") to aggregations, too.
